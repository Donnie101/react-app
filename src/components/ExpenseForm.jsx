import React, { Component } from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

const now = moment();

class ExpenseForm extends Component {

    constructor(props) {
        super(props);
        this.state = {
            description:props.expense?props.expense.description: '',
            note: props.expense?props.expense.note: '',
            amount: props.expense?(props.expense.amount/100).toString(): '',
            createdAt: props.expense?moment(props.expense.createdAt): '',
            calendarFocused: false,
            error: ''
        }
    }

    onDescripionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }))
    }

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }))
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }))
        }
    }

    onDateChange = (createdAt) => {
        if (createdAt)
            this.setState(() => ({ createdAt }))
    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }))
    }

    handleSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'PLEASE PROVIDE INFO' }))
        } else {
            this.setState(() => ({ error: '' }));
            const { description, amount, createdAt, note } = this.state;
            this.props.onSubmit({
                description,
                amount: parseFloat(amount, 10) * 100,
                createdAt: createdAt.valueOf(),
                note
            })
        }
    }

    render() {
        return (
            <div>
                {this.state.error || <p>{this.state.error}</p>}
                <form onSubmit={this.handleSubmit} action="">
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescripionChange}
                    />
                    <input
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                        type="text"
                        placeholder="amount"
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        value={this.state.note}
                        onChange={this.onNoteChange}
                        placeholder="Add a note">
                    </textarea>
                    <button type="submit">Add Expense</button>
                </form>
            </div>
        );
    }
}

export default ExpenseForm;