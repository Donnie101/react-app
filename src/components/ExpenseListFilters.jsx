import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates'
import { setEndDate,setStartDate, setTextFilter, sortByAmount, sortByDate } from '../actions/filters'


class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    }

    onDatesChange = ({startDate,endDate})=>{
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate)
    }

    onFocusChange = (calendarFocused)=>{
        this.setState(()=>({calendarFocused}))
    }

    render() {
        return (
            <div>
                <input type="text" onChange={(e) => {
                    setTextFilter(e.target.value)
                }} value={this.props.filters.text} />
                <select
                    value={this.props.filters.sortBy}
                    onChange={(e) => {
                        e.target.value === 'date' ? this.props.sortByDate() : this.props.sortByAmount();
                    }}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={()=>false}
                    showClearDates={true}
                />
            </div>
        );
    }
}

function mapStateToProps({ filters }) {
    return { filters }
}

export default connect(mapStateToProps,
    {
        setTextFilter,
        sortByAmount,
        sortByDate,
        setStartDate,
        setEndDate
    }
)(ExpenseListFilters);