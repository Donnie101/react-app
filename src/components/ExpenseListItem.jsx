import React from 'react';
import {connect} from 'react-redux'
import {Link} from 'react-router-dom'
import {removeExpense} from '../actions/expenses';

const ExpenseListItme = ({ expense: {id, description, amount, createdAt },removeExpense }) => {
    return (
        <div>
            <h3>{description}</h3>
            <p>{amount} - {createdAt}</p>
            <Link to={`/edit/${id}`}>Edit</Link>
        </div>
    )
}

export default connect(null,{removeExpense})(ExpenseListItme);
