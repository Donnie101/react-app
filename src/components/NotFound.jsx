import React, { Component } from 'react';
import { Link } from 'react-router-dom';

const FourOhFour = () => (
    <div>
        <p>Page Not Found - 404</p>
        <Link to="/">Go Home</Link>
    </div>
);

export default FourOhFour;