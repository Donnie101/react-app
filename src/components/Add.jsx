import React from 'react';
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux'
import {Redirect} from 'react-router-dom';

import {addExpense} from '../actions/expenses';

const Add = ({addExpense,history:{push}}) => {
    return (
        <div>
            <h1>Add Expense</h1>
            <ExpenseForm 
                onSubmit={(expense)=>{
                    addExpense(expense);
                    push('/');
                }}
            />
        </div>
    ); 
}

export default connect(null,{addExpense})(Add);
