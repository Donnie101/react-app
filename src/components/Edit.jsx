import React, { Component } from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

const Edit = ({ expense, editExpense, removeExpense, history: { push } }) => {
    return (
        <div>
            <h1>Edit Expense</h1>
            <ExpenseForm
                expense={expense}
                onSubmit={(updates) => {
                    editExpense(expense.id, updates)
                    push('/');
                }}
            />
            <button onClick={() => {
                removeExpense(expense.id)
                push('/');
            }}>Remove Expense</button>
        </div>
    );
}

function mapStateToProps({ expenses }, { match: { params: { id } } }) {
    return {
        expense: expenses.find((expense) => (id === expense.id))
    }
}

export default connect(mapStateToProps, { editExpense, removeExpense })(Edit);