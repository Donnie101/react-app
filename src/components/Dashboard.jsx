import React, { Component } from 'react';
import ExpenseList from './ExpeseList'
import ExpenseListFilters from './ExpenseListFilters';

const Dashboard = () => {
    return (
        <div>
            <ExpenseListFilters/>
            <ExpenseList/>
        </div>
    );
}

export default Dashboard;