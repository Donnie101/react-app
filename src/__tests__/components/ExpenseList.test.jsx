import React from 'react';
import { shallow } from 'enzyme';
import {ExpenseList} from '../../components/ExpeseList';
import expenses from '../../fixtuers/expenses';

test('render ExpenseList',()=>{
    const wrapper = shallow(<ExpenseList expenses={expenses}/>)
    expect(wrapper).toMatchSnapshot()
});