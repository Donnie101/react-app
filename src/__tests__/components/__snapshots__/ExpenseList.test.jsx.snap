// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render ExpenseList 1`] = `
<div>
  <Connect(ExpenseListItme)
    expense={
      Object {
        "amount": 195,
        "createdAt": 0,
        "description": "Gum",
        "id": "1",
        "note": "",
      }
    }
    key="1"
  />
  <Connect(ExpenseListItme)
    expense={
      Object {
        "amount": 109500,
        "createdAt": -345600000,
        "description": "Rent",
        "id": "2",
        "note": "",
      }
    }
    key="2"
  />
  <Connect(ExpenseListItme)
    expense={
      Object {
        "amount": 4500,
        "createdAt": 345600000,
        "description": "Credit Card",
        "id": "3",
        "note": "",
      }
    }
    key="3"
  />
</div>
`;
